glean-core/ffi/src/upload.rs

```diff
@@ -0,0 +1,90 @@
+use std::ffi::{CStr, CString};
+use std::os::raw::c_char;
+
+use ffi_support::IntoFfi;
+
+use crate::glean_str_free;
+use glean_core::{upload::PingUploadTask, Result};
+
+/// A FFI-compatible representation for the PingUploadTask
+#[repr(u8)]
+pub enum FfiPingUploadTask {
+    Wait,
+    Upload {
+        uuid: *mut c_char,
+        path: *mut c_char,
+        body: *mut c_char,
+        headers: *mut c_char,
+    },
+    Done,
+}
+
+impl FfiPingUploadTask {
+    pub fn uuid_as_str(&self) -> Result<&str> {
+        if let FfiPingUploadTask::Upload { uuid, .. } = self {
+            assert!(!uuid.is_null());
+            unsafe {
+                CStr::from_ptr(*uuid)
+                    .to_str()
+                    .map_err(|_| glean_core::Error::utf8_error())
+            }
+        } else {
+            Ok("")
```

Comment:
Is it common practice in Rust to return empty results? Shouldn't be return value be something like `Result<(), &str>`? Or what are you trying to do here?

Reply:
Indeed, this seems like it should be an error (if there's no `uuid` we can't return it)

Reply:
Which brings up the question if that should not just be in the one place where it is used directly, instead of making it a function (within `glean_process_ping_upload_response`)

Reply:
I actually was not sure which error to return here, forgot to comment it. None of the errors in our error module seemed to fit. Should I add a new one?

Reply:
Sure, not making it a separate function eliminates that problem. I'll do that.